File: Python/BOJ/9466.py
import sys
readl = sys.stdin.readline
sys.setrecursionlimit(10**6)

def dfs(num):
  global count
  
  visited[num] = True
  team.append(num)

  pick = arr[num]
  if not visited[pick]:
    dfs(pick)
  else:
    if pick in team:
      count += len(team[team.index(pick):])
      
t = int(readl())
for _ in range(t):
  n = int(readl())
  arr = [0] + list(map(int, readl().split()))
  
  visited = [False for _ in range(n + 1)]
  count = 0
  for i in range(1, n+1):
    if not visited[i]:
      team = []
      dfs(i)

  result = n - count
  print(result)

File: Swift/BOJ/5000/9466.swift
//
//  9466.swift
//  SwiftAlgorithm
//
//  Created by 유정주 on 2023/04/12.
//

import Foundation

//사이클이 만들어지면 같은 팀
//n - 팀 인원 개수

func dfs(_ num: Int) {
    visited[num] = true
    team.append(num)
    
    let pick = input[num]
    if !visited[pick] {
        dfs(pick)
    } else {
        if let firstIndex = team.firstIndex(of: pick) {
            count += Array(team[firstIndex...]).count
        }
    }
}

let testcase = Int(readLine()!)!
var n = 0
var input: [Int] = []
var visited: [Bool] = []
var team: [Int] = []
var count = 0
for _ in 0..<testcase {
    n = Int(readLine()!)!
    input = readLine()!.split { $0 == " " }.map { Int($0)! - 1 }
    visited = Array(repeating: false, count: n)
    count = 0
    for i in 0..<n {
        if !visited[i] {
            team = []
            dfs(i)
        }
    }
    
    let result = n - count
    print(result)
}
